modules/dxp/apps/commerce/commerce-machine-learning-impl/src/main/java/com/liferay/commerce/machine/learning/internal/recommendation/data/source/FrequentPatternCommerceMLRecommendationCPDataSourceImpl.java

```diff
@@ -0,0 +1,150 @@
+ * distribution rights of the Software.
+ *
+ *
+ *
+ */
+
+package com.liferay.commerce.machine.learning.internal.recommendation.data.source;
+
+import com.liferay.commerce.account.model.CommerceAccount;
+import com.liferay.commerce.machine.learning.recommendation.FrequentPatternCommerceMLRecommendation;
+import com.liferay.commerce.machine.learning.recommendation.FrequentPatternCommerceMLRecommendationManager;
+import com.liferay.commerce.product.catalog.CPCatalogEntry;
+import com.liferay.commerce.product.constants.CPWebKeys;
+import com.liferay.commerce.product.data.source.CPDataSource;
+import com.liferay.commerce.product.data.source.CPDataSourceResult;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.language.LanguageUtil;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.util.ListUtil;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Riccardo Ferrari
+ */
+@Component(
+	enabled = false, immediate = true,
+	property = "commerce.product.data.source.name=" + FrequentPatternCommerceMLRecommendationCPDataSourceImpl.NAME,
+	service = CPDataSource.class
+)
+public class FrequentPatternCommerceMLRecommendationCPDataSourceImpl
+	extends BaseCommerceMLRecommendationCPDataSource {
+
+	public static final String NAME =
+		"frequentPatternCommerceMLRecommendationDataSource";
+
+	@Override
+	public String getLabel(Locale locale) {
+		return LanguageUtil.get(
+			getResourceBundle(locale), "frequent-pattern-recommendations");
+	}
+
+	@Override
+	public String getName() {
+		return NAME;
+	}
+
+	@Override
+	public CPDataSourceResult getResult(
+			HttpServletRequest httpServletRequest, int start, int end)
+		throws Exception {
+
+		CommerceAccount commerceAccount =
+			commerceAccountHelper.getCurrentCommerceAccount(httpServletRequest);
+
+		if (commerceAccount == null) {
+			return new CPDataSourceResult(Collections.emptyList(), 0);
+		}
+
+		CPCatalogEntry cpCatalogEntry =
+			(CPCatalogEntry)httpServletRequest.getAttribute(
+				CPWebKeys.CP_CATALOG_ENTRY);
+
+		if (cpCatalogEntry == null) {
+			return new CPDataSourceResult(Collections.emptyList(), 0);
+		}
+
+		List<FrequentPatternCommerceMLRecommendation>
+			frequentPatternCommerceMLRecommendations =
+				_frequentPatternCommerceMLRecommendationManager.
+					getFrequentPatternCommerceMLRecommendations(
+						portal.getCompanyId(httpServletRequest),
+						new long[] {cpCatalogEntry.getCPDefinitionId()});
+
+		if (frequentPatternCommerceMLRecommendations.isEmpty()) {
+			return new CPDataSourceResult(Collections.emptyList(), 0);
+		}
+
+		long groupId = portal.getScopeGroupId(httpServletRequest);
```

Comment:
We can move `long groupId = portal.getScopeGroupId(httpServletRequest);` inside the try block, since its only being used in there.

Reply:
done